Overhaul CI documentation (#183)
freedomofpress/securedrop-dev-docs: docs/continuous_integration.rst

.. _ci_tests:

Continuous Integration
===============================

The SecureDrop project uses `GitHub Actions <https://docs.github.com/en/actions>`_ for
running automated continuous integration on code changes. You can get an overview of what

> each project does by reviewing the ``Makefile`` and files in the ``.github/workflows`` folder.

Reviewer: I initially was a bit confused by which makefile and workflows folder this was referring to. I think the below suggestion is accurate, but let me know if I overlooked something.

```suggestion
each project does by reviewing the ``Makefile`` and files in the ``.github/workflows`` folder of the project's repository.
```

Basics
------

Our CI runs a mixture of linters, tests and build processes to validate code submissions.

Most tasks have a corresponding ``make`` target that will run the same command locally. Common
targets across all our projects include:

* ``make lint``: run linters
* ``make fix``: apply automated fixes from formatters and linters
* ``make test``: run automated tests

In CI, these are run in a container using the corresponding Linux distribution (e.g. Debian or Fedora),

> which can also be used to reproduce CI results locally. Some projects, like ``securedrop`` (server) and

Reviewer: Are these make targets used directly in the CI itself, or do they replicate the CI commands? If they are maintained separately, is it worth calling that out just in case there's a desync in the future and the CI ends up running different commands than the makefile?

Author: In nearly all cases we run the make targets in CI itself.

``securedrop-workstation``, automatically run commands in containers.

Pull requests
-------------

> Most CI jobs are triggered by both ``push`` and ``pull_request`` events. The former is run against

Reviewer: Is it worth linking to the github docs for the [`push`](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#push) and [`pull_request`](https://github.com/freedomofpress/infrastructure/issues/4984#issuecomment-2305662370) events? For users not familiar with GHA those might be helpful references, but I'm not sure if it's necessarily worth including.

Author: Good idea, done.

your branch, while the latter is run against your branch merged into ``main`` (or ``develop``).

Special branch prefixes
-----------------------

In the ``securedrop`` repository, some slower jobs are only triggered if a specific branch prefix
is used when creating the pull request. Currently these are:

* ``stg-*``: runs a staging build in GCE, see :ref:`Configuration Tests<config_tests>`
* ``l10n-*``: runs localization tests across all 20+ supported languages

Nightlies
---------

> For ``securedrop-workstation`` and ``securedrop-client``, we build packages on every merged
> commit and every night and publish them to our test yum and apt repositories respectively.

Reviewer: I think this is a bit more readable, but it's mostly stylistic preferences

```suggestion
For ``securedrop-workstation`` and ``securedrop-client``, a package is built for every merged
commit as well as every night. These packages are published to the test yum and apt repositories.
```

Author: I like this, will tweak slightly (we build multiple packages from client repo).

> commit and every night and publish them to our test yum and apt repositories respectively.

Reviewer: I'm not sure I understand the "respectively" in this sentence. Is it saying that the per-commit builds are pushed to yum and the nightly builds are pushed to apt? Or that both the per-commit and nightly builds build yum and apt artifacts which are then pushed to their respective repositories?

Author: I was trying to say that workstation -> rpm and client -> apt, but I don't think it actually helped, given the confusion!

> commit and every night and publish them to our test yum and apt repositories respectively.

Reviewer: Is it worth specifying when "every night" is? Not sure if that's the kind of detail worth putting in the docs or if people want to know they can go lookup the workflow.

Author: Yeah, I think if people want to know the exact time they can just look it up.

A "nightlies" workflow runs in each repository that builds the respective packages. The workflow
uses an authenticated token for the ``sdcibot`` GitHub account to push the packages and build metadata
to ``build-logs``, ``securedrop-apt-test`` and ``securedrop-yum-test``.

The ``securedrop-apt-test`` and ``securedrop-yum-test`` repositories have a workflow that automatically

> prunes older packages, also using a token for ``sdcibot`` to push to themselves.

Reviewer: In general for this document, is it worth directly linking to the repositories themselves (i.e. https://github.com/freedomofpress/securedrop-apt-test)? That may help users more easily navigate/investigate this process if they're trying to understand the different components/

Author: Really I need to finish https://github.com/freedomofpress/securedrop-dev-docs/pull/81 and then cross-link to those docs.
